File: leetcode/search/611. Valid Triangle Number/py/main.py
import bisect


class Solution:
    def triangleNumber(self, nums):
        n = len(nums)
        ans = 0
        nums.sort()

        # The main condition nums[i] + nums[j] > nums[k] where i < j < k

        for i in range(n):
            for j in range(i + 1, n):
                # left leftmost index which nums[left_most_index] >= nums[i] + nums[j]
                k = bisect.bisect_left(nums, nums[i] + nums[j])

                ans += max(0, k - j - 1)

        return ans

File: leetcode/search/611. Valid Triangle Number/js/main.js
// Origin idea https://leetcode.com/problems/valid-triangle-number/discuss/128135/A-similar-O(n2)-solution-to-3-Sum

// Three pointers (from the right to the left)

// Time O(N^2)
// Space O(1)
const triangleNumber = nums => {
  let ans = 0;
  let n = nums.length;
  nums.sort((a, b) => a - b);

  for (let fixedRightPointer = n - 1; fixedRightPointer > 1; fixedRightPointer--) {
    // fixedRightPointer - it's a fixed right border

    let start = 0;
    let end = i - 1;

    // Scan the range [start, end] and inc start or dec right window's border
    // When we complite the scan inc fixedRightPointer and start new range search
    while (start < end) {
      if (nums[start] + nums[end] > nums[fixedRightPointer]) {
        ans += end - start;
        end -= 1;
      } else {
        start += 1;
      }
    }
  }

  return ans;
};
